06_predict_survival/nbconverted/plot_survival_curves.py

```diff
@@ -0,0 +1,209 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ### Plot survival curves
+# 
+# Just for BRCA now, as a sanity check.
+
+# In[1]:
+
+
+import os
+import pickle as pkl
+
+import numpy as np
+import pandas as pd
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+import mpmp.config as cfg
+import mpmp.utilities.analysis_utilities as au
+import mpmp.utilities.data_utilities as du
+
+get_ipython().run_line_magic('load_ext', 'autoreload')
+get_ipython().run_line_magic('autoreload', '2')
+
+
+# In[2]:
+
+
+cancer_type = 'BRCA'
+signal = 'signal'
+
+functions_dir = os.path.join(cfg.repo_root, 'test_results_10')
+samples, fns = au.load_survival_curves(functions_dir, cancer_type, signal=signal)
+
+print(len(samples), len(fns))
+
+
+# In[3]:
+
+
+# get clinical info for BRCA samples
+# this includes age, subtype, etc.
+clinical_df = pd.read_excel(cfg.clinical_data,
+                            sheet_name='TCGA-CDR',
+                            index_col='bcr_patient_barcode',
+                            engine='openpyxl')
+
+clinical_df.drop(labels=['Unnamed: 0'], axis=1, inplace=True)
+clinical_df.rename(columns={'age_at_initial_pathologic_diagnosis': 'age'}, inplace=True)
+
+print(clinical_df.shape)
+clinical_df.head()
+
+
+# ### Color survival curves by age
+# 
+# Here, we discretize age into quantiles, and color the survival curves by quantile. We'd expect older patients to have shorter/poorer survival in general.
+
+# In[4]:
+
+
+ages = clinical_df.loc[(clinical_df.type == cancer_type) & (~clinical_df.age.isna()), 'age']
+print(ages.isna().sum())
+ages.head()
+
+
+# In[5]:
+
+
+sns.set()
+sns.histplot(ages)
+qs_age = []
+for q in [0.8]:
+    q_age = np.quantile(ages, q)
+    qs_age.append(q_age)
+    plt.gca().axvline(q_age, color='red', linestyle='--', linewidth=3)
+
+
+# In[6]:
+
+
+age_quantiles = np.digitize(ages.values, qs_age)
+ages = pd.DataFrame(ages)
+ages['age_quantile'] = age_quantiles
+ages.head(n=10)
+
+
+# In[7]:
+
+
+sns.set({'figure.figsize': (10, 8)})
+
+# color by age binned into quantiles
+for ix in range(len(fns)):
+    try:
+        fn = fns[ix]
+        sample_id = '-'.join(samples[ix].split('-')[:3])
+        sample_age_quantile = int(ages.loc[sample_id, 'age_quantile'])
+        plt.step(fn.x, fn.y, label=sample_age_quantile,
+                 color='C{:d}'.format(sample_age_quantile))
+    except KeyError:
+        continue
+    
+handles, labels = plt.gca().get_legend_handles_labels()
+by_label = dict(zip(labels, handles))
+plt.legend(by_label.values(), by_label.keys(), title='Above/below 80th age percentile')
+plt.xlabel('Time')
+plt.ylabel('Survival probability')
```

Comment:
Each line is one patient. With survival prediction, rather than predicting a single value for each sample (like you would in classification or regression) you predict a function which gives the survival probability at each time point.

So for these plots, I'm doing 4-fold cross-validation on the dataset, and for each held-out patient I'm plotting their predicted survival function and coloring the predicted functions by covariates that we think might correlate with survival.

[More detail here](https://scikit-survival.readthedocs.io/en/stable/user_guide/00-introduction.html#The-Survival-Function) if you're interested.